### PR TITLE
PHOENIX-5066 The TimeZone is incorrectly used during writing or reading data

### PR DESCRIPTION
**Pull request by richardantal**

### CODE REVIEW

File: phoenix-core/src/main/java/org/apache/phoenix/expression/function/RoundJodaDateExpression.java [hunk 1]
```diff
@@ -48,9 +49,10 @@ public boolean evaluate(Tuple tuple, ImmutableBytesWritable ptr) {
             }
             PDataType dataType = getDataType();
             long time = dataType.getCodec().decodeLong(ptr, children.get(0).getSortOrder());
-            DateTime dt = new DateTime(time,ISOChronology.getInstanceUTC());
+            int offset = Calendar.getInstance().getTimeZone().getRawOffset();
+            DateTime dt = new DateTime(time + offset,ISOChronology.getInstanceUTC());
             long value = roundDateTime(dt);
-            Date d = new Date(value);
+            Date d = new Date(value - offset);
```
Comment (dbwong) [inline phoenix-core/src/main/java/org/apache/phoenix/expression/function/RoundJodaDateExpression.java:55]: In RoundDate expression offset was handled inside of the rounding function.  I'd either recommend the same pattern or maybe handling this in a utility class?

File: phoenix-core/src/main/java/org/apache/phoenix/jdbc/PhoenixConnection.java [hunk 2]
```diff
@@ -331,11 +339,16 @@ public ReadOnlyProps getProps() {
...
                 QueryServices.MAX_MUTATION_SIZE_BYTES_ATTRIB,
                 QueryServicesOptions.DEFAULT_MAX_MUTATION_SIZE_BYTES);
-        String timeZoneID = this.services.getProps().get(QueryServices.DATE_FORMAT_TIMEZONE_ATTRIB,
-                DateUtil.DEFAULT_TIME_ZONE_ID);
-        Format dateFormat = DateUtil.getDateFormatter(datePattern, timeZoneID);
-        Format timeFormat = DateUtil.getDateFormatter(timePattern, timeZoneID);
-        Format timestampFormat = DateUtil.getDateFormatter(timestampPattern, timeZoneID);
+
+        String timeZoneID = this.services.getProps().get(
+                QueryServices.DATE_FORMAT_TIMEZONE_ATTRIB, DateUtil.DEFAULT_TIME_ZONE_ID);
```
Comment (dbwong) [inline phoenix-core/src/main/java/org/apache/phoenix/jdbc/PhoenixConnection.java:344]: I saw the discussion that jdbc differs from phoenix on default timezone being local vs phoenix's gmt, we should likely document in the code a bit here.

File: phoenix-core/src/main/java/org/apache/phoenix/jdbc/PhoenixPreparedStatement.java [hunk 3]
```diff
@@ -491,15 +517,39 @@ public void setString(int parameterIndex, String x) throws SQLException {
...
-            x = new Time(x.getTime());
+            String tz = this.connection.getQueryServices().getProps()
+                    .get(QueryServices.DATE_FORMAT_TIMEZONE_ATTRIB, "GMT");
+            int offset = TimeZone.getTimeZone(tz).getRawOffset();
+
+            TimeZone timezone = Calendar.getInstance().getTimeZone();
+            long msFromEpochGmt = x.getTime();
+            int offsetFromUTC = timezone.getOffset(msFromEpochGmt);
+
+            Calendar gmtCal = Calendar.getInstance(TimeZone.getTimeZone(tz));
```
Comment (dbwong) [inline phoenix-core/src/main/java/org/apache/phoenix/jdbc/PhoenixPreparedStatement.java:548]: Looks like similar code again thoughts on utility function?

File: phoenix-core/src/main/java/org/apache/phoenix/jdbc/PhoenixConnection.java [hunk 4]
```diff
@@ -171,6 +172,13 @@
     private boolean isRunningUpgrade;
     private LogLevel logLevel;
     private Double logSamplingRate;
+    private static ThreadLocal<DateUtil> DateUtilContext
```
Comment (joshelser) [inline phoenix-core/src/main/java/org/apache/phoenix/jdbc/PhoenixConnection.java:175]: Should be final? Follow code style of all caps for static+final variables, too.

**Comment by yanxinyi** [comment]
@richardantal can you rebase and handle all the suggestions, thanks

**Comment by richardantal** [comment]
I stopped working on this issue because there were lot of failing tests locally (outside of GMT timeZone) because of this change and I did not have the time to fix them.
Also after [Istvan's comment ](https://issues.apache.org/jira/browse/PHOENIX-5066?focusedCommentId=17216968&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17216968) I am not sure if it is worth to continue this solution.

**Pull request closed without merging**